boba/lang.py

```diff
@@ -4,57 +4,65 @@
 PY = 'python'
 R = 'r'
 
+script = '{{script_name}}'
+compiled = '{{universe_name}}'
+
+DEFAULT_LANGS = {
+    'python' : {
+        'ext' : ['py'],
+        'run' : ['python', script]
+    },
+    'r' : {
+        'ext' : ['R', 'r', 'rmd'],
+        'run' : ['Rscript', script]
+    }
+}
 
 class LangError(NameError):
     pass
 
 
 class Lang:
-    def __init__(self, lang, script):
-        self.lang = self._infer_lang(lang, script)
-
-    def _is_valid_lang(self, lang):
-        return lang == PY or lang == R
-
-    def _ext_to_lang(self, ext):
-        ext = ext.lower()[1:]
-        if ext == 'py':
-            return PY
-        if ext == 'r' or ext == 'rmd':
-            return R
-        return ''
-
-    def _infer_lang(self, lang, script):
-        lang = lang.strip().lower()
-
-        if self._is_valid_lang(lang):
-            return lang
-
-        if lang == '':
-            _, ext = os.path.splitext(script)
-            lang = self._ext_to_lang(ext)
-            if self._is_valid_lang(lang):
-                return lang
-            else:
-                raise LangError('Error: cannot infer language from file ' +
-                                'extension {}'.format(ext))
+    def __init__(self, script, lang=None, supported_langs=None):
+        if supported_langs:
+            self.supported_langs = supported_langs
```

Comment:
should be a combination of supported lang and default lang ... specifically, supported langs can override default lang, but otherwise anything in default lang is still available